### PR TITLE
feat: [CCM-5386]: Anomaly alert feature

### PR DESCRIPTION
**Pull request by jyotiarora-harness**
##### Summary:

Anomalies notification alerts functionality

##### Jira Links:

https://harness.atlassian.net/browse/CCM-5386

##### Screenshots:

Screenshot 2022-03-23 at 11 00 57 AM
Screenshot 2022-03-23 at 11 01 05 AM

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

**Comment by jyotiarora-harness** [comment]
retrigger titlecheck

**Comment by jyotiarora-harness** [comment]
trigger cypress

**Comment by jyotiarora-harness** [comment]
trigger cypress

**Comment by jyotiarora-harness** [comment]
trigger cypress

**Comment by jyotiarora-harness** [comment]
trigger cypress

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    borderLeft: 0,
+    paddingBottom: 0,
+    overflow: 'hidden'
+  }
+}
+
+interface AlertDialogProps {
+  hideAnomaliesAlertModal: any
+  handleSubmit: any
+  notificationData: any
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:38]: Avoid using `any`

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+          alertList: channelsData || []
+        }}
+        render={formikProps => {
+          return (
+            <FormikForm>
+              <StepWizard
+                icon="right-bar-notification"
+                iconProps={{
+                  size: 34,
+                  color: 'white'
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:87]: knit - `Color.White`

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  const { mutate: updateNotificationAlert } = useUpdateNotificationSetting({
+    queryParams: {
+      accountIdentifier: accountId,
+      perspectiveId: ''
+    }
+  })
+
+  /* istanbul ignore next */
+  const handleSubmit = async (data: any) => {
+    const payload = data.alertList.map((item: any) => {
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:132]: avoid `any`

File: src/modules/75-ce/components/AnomaliesAlert/PerspectiveSelection.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    <React.Fragment>
+      <Icon name="cross" size={16} onClick={() => props.onClose()} className={css.closeBtn} />
+      <Layout.Vertical spacing="xxlarge">
+        <Text color={Color.BLACK} font={{ variation: FontVariation.H5 }}>
+          {props.name}
+        </Text>
+        <FormInput.Select
+          items={props.items}
+          className={css.perspectiveSelection}
+          name={'perspective'}
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesAlert/PerspectiveSelection.tsx:47]: move to i18n

File: src/modules/75-ce/components/AnomaliesSearch/AnomaliesSearch.tsx [hunk 5]
```diff
@@ -8,17 +8,15 @@
 import React from 'react'
 import { Button, ButtonSize, ButtonVariation, Layout, ExpandingSearchInput } from '@wings-software/uicore'
 import { useStrings } from 'framework/strings'
-import { useFeatureFlag } from '@common/hooks/useFeatureFlag'
-import { FeatureFlag } from '@common/featureFlags'
 import css from '../../pages/anomalies-overview/AnomaliesOverviewPage.module.scss'
 
 interface SearchProps {
   onChange: React.Dispatch<React.SetStateAction<string>>
+  showModal: any
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesSearch/AnomaliesSearch.tsx:15]: `any`

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+            padding="large"
+          >
+            {getString('ce.anomalyDetection.settings.options.header')}
+          </Text>
+          <Icon name="cross" size={16} color={Color.WHITE} onClick={() => hideDrawer()} data-testid="closeDrawerIcon" />
+        </Layout.Horizontal>
+        <ul className={css.listingOptions}>
+          <li className={cx(css.listOptionItem, activePanelId === 1 && css.listOptionItemSelected)}>
+            <Text
+              font={{ size: 'normal', weight: 'semi-bold' }}
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx:279]: `FontVariation`

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    setActivePanelId(id)
+  }
+
+  return (
+    <Layout.Horizontal className={css.container}>
+      <AlertsSection />
+      <Container className={css.settingsDrawer} background={Color.PRIMARY_8}>
+        <Layout.Horizontal className={css.settingsLabelWarpper}>
+          <Text
+            font={{ size: 'normal', weight: 'semi-bold' }}
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx:267]: `FontVariation`

File: src/modules/75-ce/components/AnomaliesSettings/__test__/AnomaliesSettings.test.tsx [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        </TestWrapper>
+      </Provider>
+    )
+
+    const spinner = container.querySelector('[data-testid="loader"]')
+    expect(spinner).toBeTruthy()
+  })
+
+  test('Should update the active tab on tabs click', async () => {
+    const hideDrawer = jest.fn()
```
Comment (jenil-harness) [inline src/modules/75-ce/components/AnomaliesSettings/__test__/AnomaliesSettings.test.tsx:217]: Knit - Can move it to top of the file, rather that defining again and again

File: src/modules/75-ce/constants.ts [hunk 9]
```diff
@@ -164,3 +164,35 @@ export const allCloudProvidersList = [
     value: 'gcp'
   }
 ]
+
+export const notificationChannelsList = [
+  {
+    label: 'Slack webhook url',
```
Comment (jenil-harness) [inline src/modules/75-ce/constants.ts:170]: Move labels to i18n

File: src/modules/75-ce/components/PerspectiveReportsAndBudget/__test__/PerspectiveReportsAndBudget.test.tsx [hunk 10]
```diff
@@ -237,3 +268,115 @@ describe('test cases for Perspective Create Budgets', () => {
...
+
+    const { container } = render(
+      <Provider value={responseState as any}>
+        <TestWrapper pathParams={params}>
+          <AnomalyAlerts />
+        </TestWrapper>
+      </Provider>
+    )
+
+    await waitFor(() => Promise.resolve())
```
Comment (jenil-harness) [inline src/modules/75-ce/components/PerspectiveReportsAndBudget/__test__/PerspectiveReportsAndBudget.test.tsx:372]: ?

File: src/modules/75-ce/components/PerspectiveReportsAndBudget/PerspectiveReportsAndBudgets.tsx [hunk 11]
```diff
@@ -564,17 +710,32 @@ const RenderEmailAddresses = ({ emailAddresses = [] }: { emailAddresses: string[
...
         alignItems: 'center'
       }}
+      className={className}
     >
-      <Icon size={14} name={'Edit'} color="primary7" onClick={onClickEdit} className={css.icon} />
-      <Icon size={14} name={'trash'} color="primary7" onClick={onClickDelete} className={css.icon} />
+      <Icon
+        size={14}
+        name={'Edit'}
+        color="primary7"
```
Comment (jenil-harness) [inline src/modules/75-ce/components/PerspectiveReportsAndBudget/PerspectiveReportsAndBudgets.tsx:726]: `Color`

**Review by jenil-harness — CHANGES_REQUESTED** [review]

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx (hunk 2 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx]: Updated

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx (hunk 3 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx]: Updated according to the data type

File: src/modules/75-ce/components/AnomaliesSearch/AnomaliesSearch.tsx (hunk 5 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/AnomaliesSearch/AnomaliesSearch.tsx]: Updated according to the data type

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx (hunk 7 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx]: Updated

File: src/modules/75-ce/components/AnomaliesSettings/__test__/AnomaliesSettings.test.tsx (hunk 8 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/AnomaliesSettings/__test__/AnomaliesSettings.test.tsx]: Moved to the top

File: src/modules/75-ce/constants.ts (hunk 9 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/constants.ts]: Done

File: src/modules/75-ce/components/PerspectiveReportsAndBudget/__test__/PerspectiveReportsAndBudget.test.tsx (hunk 10 above)
Comment (jyotiarora-harness) [reply src/modules/75-ce/components/PerspectiveReportsAndBudget/__test__/PerspectiveReportsAndBudget.test.tsx]: It's unused, so removed

**Review by jenil-harness — APPROVED** [review]

**Review by rajarshee-harness — APPROVED** [review]

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+  notificationData
+}) => {
+  const { getString } = useStrings()
+
+  const [{ data: perspectiveData }] = useFetchPerspectiveListQuery()
+
+  /* istanbul ignore next */
+  const perspectiveList = (perspectiveData?.perspectives?.customerViews || []) as QlceView[]
+
+  const items = perspectiveList.map(pName => ({
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:71]: we can memoize

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    label: pName.name as string,
+    value: pName.id as string
+  }))
+
+  const channelsData =
+    notificationData?.channels?.map((item: CCMNotificationChannel) => {
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:77]: we can memoize

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  /* istanbul ignore next */
+  const handleSubmit = async (data: FormValues) => {
+    const payload = data.alertList.map((item: AlertsData) => {
+      const channel = item.channelName
+
+      if (channel === 'EMAIL' && typeof item.channelUrl === 'string') {
+        const emailList = item.channelUrl.split(',')
+        return {
+          type: channel,
+          [channelNameUrlMapping[channel as keyof typeof channelNameUrlMapping]]: emailList
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:154]: we can extract this out in a variable and use in line 154 and line 159

File: src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+
+    const queryParams = {
+      perspectiveId: data.perspective,
+      accountIdentifier: accountId
+    }
+
+    try {
+      let response
+      if (props.selectedAlert && props.selectedAlert.channels?.length) {
+        response = await updateNotificationAlert({ channels: payload as CCMNotificationChannel[] }, { queryParams })
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesAlert/AnomaliesAlertDialog.tsx:171]: we aren't we using the hooks for making the API calls

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
+  useEffect(() => {
+    if (selectedAlert && selectedAlert.perspectiveId) {
+      openAnomaliesAlertModal()
+    }
+  }, [openAnomaliesAlertModal, selectedAlert])
+
+  const deleteNotification = async (perspectiveId: string) => {
+    try {
+      const response = await deleteNotificationAlert(void 0, {
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx:88]: please use hooks wherever possible

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+    } catch (error) {
+      showError(getErrorInfoFromErrorObject(error))
+    }
+  }
+
+  const onEdit = (notificationData: CCMNotificationSetting) => {
+    setSelectedAlert(notificationData)
+  }
+
+  const actionCell: Renderer<CellProps<CCMNotificationSetting>> = ({ row }) => {
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx:108]: always better to memoize

File: src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+        Cell: actionCell,
+        width: '8%'
+      }
+    ],
+    []
+  )
+
+  const renderLoader = (): JSX.Element => {
+    return (
+      <Container className={css.loader} data-testid="loader">
```
Comment (chiragp1212) [inline src/modules/75-ce/components/AnomaliesSettings/AnomaliesSettings.tsx:218]: we generally use <PageSpinner /> for loading interactions

**Review by chiragp1212 — APPROVED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

**Pull request merged**